File: reference-apps/fastapi/app/services/vault.py
"""
Vault client service for fetching secrets

Demonstrates how to integrate with HashiCorp Vault to fetch credentials
for other infrastructure services.
"""

import httpx
import logging
from typing import Optional, Dict, Any

from app.config import settings

logger = logging.getLogger(__name__)


class VaultClient:
    """Client for interacting with HashiCorp Vault"""

    def __init__(self):
        self.vault_addr = settings.VAULT_ADDR
        self.vault_token = settings.VAULT_TOKEN
        self.headers = {"X-Vault-Token": self.vault_token}

    async def get_secret(self, path: str, key: Optional[str] = None) -> Dict[str, Any]:
        """
        Fetch a secret from Vault KV v2 secrets engine

        Args:
            path: Secret path (e.g., 'postgres', 'mysql')
            key: Optional specific key to extract

        Returns:
            Secret data or specific key value
        """
        url = f"{self.vault_addr}/v1/secret/data/{path}"

        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(url, headers=self.headers, timeout=5.0)
                response.raise_for_status()

                data = response.json()
                secret_data = data.get("data", {}).get("data", {})

                if key:
                    return {key: secret_data.get(key)}

                return secret_data

        except httpx.HTTPError as e:
            logger.error(f"Failed to fetch secret from Vault: {e}")
            raise
        except Exception as e:
            logger.error(f"Unexpected error fetching secret: {e}")
            raise

    async def check_health(self) -> Dict[str, Any]:
        """Check Vault health status"""
        url = f"{self.vault_addr}/v1/sys/health"

        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{url}?standbyok=true",
                    timeout=5.0
                )

                return {
                    "status": "healthy" if response.status_code == 200 else "unhealthy",
                    "initialized": response.status_code != 501,
                    "sealed": response.status_code == 503,
                    "standby": response.status_code == 429,
                }
        except Exception as e:
            logger.error(f"Vault health check failed: {e}")
            return {
                "status": "unhealthy",
                "error": str(e)
            }


# Global Vault client instance
vault_client = VaultClient()

File: reference-apps/golang/internal/services/vault.go
package services

import (
	"context"
	"fmt"

	vault "github.com/hashicorp/vault/api"
)

// VaultClient wraps the Vault API client
type VaultClient struct {
	client *vault.Client
}

// NewVaultClient creates a new Vault client
func NewVaultClient(addr, token string) (*VaultClient, error) {
	config := vault.DefaultConfig()
	config.Address = addr

	client, err := vault.NewClient(config)
	if err != nil {
		return nil, fmt.Errorf("failed to create Vault client: %w", err)
	}

	client.SetToken(token)

	return &VaultClient{client: client}, nil
}

// GetSecret retrieves a secret from Vault KV v2
func (v *VaultClient) GetSecret(ctx context.Context, path string) (map[string]interface{}, error) {
	secret, err := v.client.KVv2("secret").Get(ctx, path)
	if err != nil {
		return nil, fmt.Errorf("failed to read secret at %s: %w", path, err)
	}

	if secret == nil || secret.Data == nil {
		return nil, fmt.Errorf("no data found at %s", path)
	}

	return secret.Data, nil
}

// GetSecretKey retrieves a specific key from a secret
func (v *VaultClient) GetSecretKey(ctx context.Context, path, key string) (interface{}, error) {
	data, err := v.GetSecret(ctx, path)
	if err != nil {
		return nil, err
	}

	value, ok := data[key]
	if !ok {
		return nil, fmt.Errorf("key %s not found in secret %s", key, path)
	}

	return value, nil
}

// HealthCheck checks if Vault is accessible and unsealed
func (v *VaultClient) HealthCheck(ctx context.Context) (map[string]interface{}, error) {
	health, err := v.client.Sys().Health()
	if err != nil {
		return nil, fmt.Errorf("failed to check Vault health: %w", err)
	}

	return map[string]interface{}{
		"initialized": health.Initialized,
		"sealed":      health.Sealed,
		"standby":     health.Standby,
		"version":     health.Version,
	}, nil
}
